server/recceiver/cfstore.py

```diff
@@ -48,125 +55,154 @@ def commit(self, TR):
         if _log.isEnabledFor(logging.DEBUG):
             _log.debug("CF_COMMIT %s", TR.infos.items())
         pvNames = [unicode(rname, "utf-8") for rid, (rname, rtype) in TR.addrec.iteritems()]
-        iocName=TR.src.port
-        hostName=TR.src.host
-    
-        '''
-        Currently using the hostIP and the iocPort
-        if 'IOCNAME' in TR.infos:
-                iocName = TR.infos['IOCNAME']
-        if 'HOSTNAME' in TR.infos:
-                hostName = TR.infos['HOSTNAME']
-        '''
+        delrec = TR.delrec
+        iocName = TR.src.port
+        hostName = TR.src.host
+        iocid = hostName + str(iocName)
         owner = TR.infos.get('CF_USERNAME') or TR.infos.get('ENGINEER') or self.conf.get('username', 'cfstore')
-        time = str(datetime.datetime.now())
-            
+        time = self.currentTime()
+        if TR.initial:
+            self.iocs[iocid] = {"iocname": iocName, "hostname": hostName, "channelcount": 0}  # add IOC to source list
+        if TR.connected:
+            for pv in pvNames:
+                if pv in self.channel_dict:
+                    if iocid not in self.channel_dict[pv]:
+                        self.channel_dict[pv].append(iocid)  # add iocname to pvName in dict
+                        self.iocs[iocid]["channelcount"] += 1
+                    # else:  # info already in dictionary, possibly recovering from ioc disconnect
+                else:
+                    self.channel_dict[pv] = [iocid]  # add pvName with [iocname] in dict
+                    self.iocs[iocid]["channelcount"] += 1
+            for pv in delrec:
+                if iocid in self.channel_dict[pv]:
+                    self.channel_dict[pv].remove(iocid)
+                    self.iocs[iocid]["channelcount"] -= 1
+                    if len(self.channel_dict[pv]) <= 0:  # case: channel has no more iocs
+                        del self.channel_dict[pv]
```

Comment:
What is "channelcount" use for?  A manual ref-counter makes me a bit nervous.  At the least, a negative ref. count should trigger a warning message.